### ISSUE TITLE
Pentest to verify TLS CVE fixes

### ISSUE THREAD

**Issue by nbolton** [deskflow/deskflow#8010]
💵 **Bounty:** $500 ([details](https://github.com/deskflow/deskflow/issues/8010#issuecomment-2562857625))

**Blocker:** 
- #7806

**Task:** Penetration Tester/Ethical Hacker/security expert wanted to pentest Deskflow and check the fixes for TLS CVEs:
- CVE-2021-42072
- CVE-2021-42073
- CVE-2021-42074
- CVE-2021-42075
- CVE-2021-42076

We have a bounty for fixing #7806. A number of PRs were opened to fix this. We want to verify these PRs do indeed solve the CVEs so that we can award the bounty.

**Comment by CyberPro010** [comment]
how can i participate ? i want payment in bitcoin or eth

**Comment by nbolton** [comment]
> how can i participate ? i want payment in bitcoin or eth

Right now, @algora-pbc does not support cryptocurrency. If nobody accepts the bounty within the next week or so, perhaps we can talk on Matrix or IRC: https://github.com/deskflow/deskflow/wiki/Chat-with-us

**Comment by ologbonowiwi** [comment]
I'm OOO for the holidays but I'd love to try this out.

What is the desired output here? code attempting to explore the mentioned vulnerabilities?

**Comment by sithlord48** [comment]
Correct we would want pen testing done to show that these are resolved. Perhaps by running the exploit on a known bad version 1.17.2 and then a suspected good version 1.18.0. Ideally this should be something we (or another 3rd party) could also try locally, Maybe even as a test run at build time to make sure we do not regress 

Please note: that  @varshith257 has a bit to finish before the first two CVEs on the list are resolved. These should be resolved with 1.18.0 
 - [CVE-2021-42074](https://github.com/advisories/GHSA-97hp-rfmm-2c7w)
 - [CVE-2021-42075](https://github.com/advisories/GHSA-c58c-q645-27mc)
 - [CVE-2021-42076](https://github.com/advisories/GHSA-g73p-h925-4x6w)

**Comment by zelosleone** [comment]
@sithlord48 @nbolton since prs got merged, is it okay to start?

**Comment by sithlord48** [comment]
YES!

**Comment by nbolton** [comment]
@mgerstner Perhaps also of interest to you.

**Comment by zelosleone** [comment]
@nbolton @sithlord48 i need some info on this. one question since i am trying out previous versions with wireshark, netcat etc. did you guys actually test out the previous versions with actual pentest env and tools instead of static code analysis?

**Comment by sithlord48** [comment]
No these cves are older and were from before the barrier fork.

**Comment by zelosleone** [comment]
> No these cves are older and were from before the barrier fork.

Jesus, none of them exists for you guys in the first place from my multiple testings. It correctly sets limits, closes ports and rejects invalid tokens easily even in 1.7 version. I was worried you guys got baited by static code analysis tools.

**Comment by sithlord48** [comment]
the server did not keep track of the connections nor does it have a authorized client list before the last PR Not all of those CVE may have been valid but CVE-2021-42072 and CVE-2021-42073 were for sure.

**Comment by zelosleone** [comment]
> the server did not keep track of the connections nor does it have a authorized client list before the last PR Not all of those CVE may have been valid but [CVE-2021-42072](https://github.com/advisories/GHSA-wvmp-5fvm-29ff?rgh-link-date=2025-02-18T13%3A01%3A13.000Z) and [CVE-2021-42073](https://github.com/advisories/GHSA-fm32-wgpg-7rc9?rgh-link-date=2025-02-18T13%3A01%3A13.000Z) were for sure.

Let me check them further, i just completed 1 hour of pentesting for file descriptor, thinking "i must be doing something wrong, they said it was there!" it was my misunderstanding i see, i will check and open a pr where you guys can manually test it as well.

**Comment by ologbonowiwi** [comment]
> Jesus, none of them exists for you guys in the first place from my multiple testings

I gave a few attempts here https://github.com/ologbonowiwi/deskflow-pentesting, primarily for CVE 42074, 42075, and 42076, and couldn't break the server for any of them. I thought I was doing something wrong, but if these vulnerabilities don't exist, that explains a lot lol

**Comment by zelosleone** [comment]
> > Jesus, none of them exists for you guys in the first place from my multiple testings
> 
> I gave a few attempts here [ologbonowiwi/deskflow-pentesting](https://github.com/ologbonowiwi/deskflow-pentesting?rgh-link-date=2025-02-18T13%3A17%3A17.000Z), primarily for CVE 42074, 42075, and 42076, and couldn't break the server for any of them. I thought I was doing something wrong, but if these vulnerabilities don't exist, that explains a lot lol

No, just broke through the server with a fake client id. @sithlord48 was right.

**Comment by zelosleone** [comment]
@sithlord48 That sonar analysis was right, and its not solved in the newer version HOWEVER even in older versions, all you can do with a fake client that is verified by deskflow is to just get verified, you cannot extract mouse-keyboard inputs at all.

**Comment by sithlord48** [comment]
Can you explain that a bit more . I am not quite understanding what CVE's your talking about.

**Comment by zelosleone** [comment]
> Can you explain that a bit more . I am not quite understanding what CVE's your talking about.

All five of them mentioned in the first post do not exist in any versions from 1.70 to 1.90. However this is still a problem https://sonarcloud.io/project/issues?open=AZH_1rPbUm1X5WafpxUI&id=deskflow_deskflow since it accepts fake client ids by dumping SYNERGY or Barrier response (synergy in 1.70-1.72 but Barrier in newer versions.) but it doesnt grant any permission to run the commands. However, it definitely accepts TLS connections with that fake client id. It also shows up as "accepted client" in deskflow logs.

**Comment by nbolton** [comment]
> However, it definitely accepts TLS connections with that fake client id. It also shows up as "accepted client" in deskflow logs.

Is this a new issue?

**Comment by zelosleone** [comment]
> > However, it definitely accepts TLS connections with that fake client id. It also shows up as "accepted client" in deskflow logs.
> 
> Is this a new issue?

I think its related to sonar report you guys shared and i quoted in my PR, it doesnt properly check and confirm client while connecting via tls:// and directly accepts the connection. However, it doesnt let you execute commands etc. So it fits with sonar report.

**Comment by nbolton** [comment]
> sonar report

But I guess we don't have an open bug report for it, or do we?

**Comment by zelosleone** [comment]
> > sonar report
> 
> But I guess we don't have an open bug report for it, or do we?

it was mentioned here: https://github.com/deskflow/deskflow/issues/7806

"[SonarCloud hotspot: cpp:S5527](https://sonarcloud.io/project/issues?open=AZH_1rPbUm1X5WafpxUI&id=deskflow_deskflow) (Enable server hostname verification on this SSL/TLS connection)"

this one.

**Comment by zelosleone** [comment]
> That's closed now so I think we need a new issue. Strange how both the SC hot spot and GitHub report auto solved if it's not actually fixed...

As a simple test, you could also try to grap the response dump after trying to connect to tls:// with a fake client id, it dumps the response and deskflow client shows the log as "accepted client" I think your approach "code" wise worked, but in practical terms it needs some filtering to host connection, but then again it only dumps the response and gets accepted, but doesnt maintain the connection enough to actually execute the commands.

**Comment by nbolton** [comment]
Fixed by: https://github.com/deskflow/scripts/pull/1

**Comment by nbolton** [comment]
@zelosleone Please `/claim` here.

**Comment by zelosleone** [comment]
@nbolton i think tipping would be better here i guess.
